### PR TITLE
Implement kserve modal

### LINKED ISSUE #1943: [Story]: KServe Deploy Modal
### Goal

Create a Modal to deploy KServe

### Dependency issue

* https://github.com/opendatahub-io/odh-dashboard/issues/1941

### Itemized goals

* Create a modal with the following fields:
   * Model name
   * Serving Runtime selection
   * Model framework selection
   *  Replicas to deploy
   * Resources
   * Accelerator support
   * Model route
   * Token authorization
   * Model location
 * Create the logic to create the `ServingRuntime` with the size and model route
 * Create the `InferenceService` linked to the `ServingRuntime`
 * The `InferenceService` **should not contain** the `'serving.kserve.io/deploymentMode': 'ModelMesh'` annotation
 * **If it's the first model deployed, add the annotation `modelmesh-enabled: false` to the project.**
 * Make sure that in modelmesh we have that same logic, with the first `ServingRuntime` deployed we add `modelmesh-enabled: true`

**This modal should have an edit prop, and should behave like the current InferenceService and ServingRuntime modals, meaning that all the fields that are disabled in editing there should be disabled here, and when a user has Edit permissions, it should behave like the ServingRutime modal**

### Aspects continued elsewhere

_No response_

### Mocks

* https://www.sketch.com/s/113593f8-5970-49d6-a352-709b07639127/a/8VAdaMa

### PR DESCRIPTION
**Pull request by lucferbux**
## Description

Closes #1943 #1945 

### Project view new creation

Screenshot 2023-10-27 at 21 33 19

1. Enabling both platforms first
2. Click the "Single model serving platform"
3. Check that the model cannot be deployed until you fill in all the values
4. Check that you get the correct serving runtimes
5. Deploy the model
6. Check that you get one Model Server and one Model deployed

### Project view with only kserve and existing data

1. In a project with "Single model platform"
2. Deploy a new model
3. Check that you get one Model Server and one Model deployed

### Global view

Screenshot 2023-10-27 at 21 32 08

1. Try to edit one model deployed in a `Single model serving enabled` project
2. Edit the project
3. Check that the changes are submitted

## Test Impact

TODO

## Request review criteria:

Self checklist (all need to be checked):
- [X] The developer has manually tested the changes and verified that the changes work
- [X] Commits have been squashed into descriptive, self-contained units of work (e.g. 'WIP' and 'Implements feedback' style messages have been removed)
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has added tests or explained why testing cannot be added (unit tests & storybook for related changes)

If you have UI changes: 

- [ ] Included any necessary screenshots or gifs if it was a UI change.
- [ ] Included tags to the UX team if it was a UI/UX change (find relevant UX in the [SMEs](https://github.com/opendatahub-io/odh-dashboard/tree/main/docs/smes.md) section).

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

### CODE REVIEW

File: frontend/src/pages/modelServing/screens/global/InferenceServiceTable.tsx [hunk 1]
```diff
@@ -82,6 +84,24 @@ const InferenceServiceTable: React.FC<InferenceServiceTableProps> = ({
...
       />
+      <ManageKServeModal
+        isOpen={!!editInferenceService && !isModelMesh(editInferenceService)}
+        editInfo={{
+          inferenceServiceEditInfo: editInferenceService,
+          servingRuntimeEditInfo: {
+            servingRuntime: servingRuntimes.find(
+              (sr) => sr.metadata.name === editInferenceService?.spec.predictor.model.runtime,
+            ),
+            secrets: [], // TODO: Get the secrets
```
Comment (lucferbux) [inline frontend/src/pages/modelServing/screens/global/InferenceServiceTable.tsx:95]: @christianvogt @DaoDaoNoCode doing a review i realized I didn't add the secrets, I need to do this later

File: frontend/src/pages/modelServing/screens/projects/ModelServingPlatform.tsx [hunk 2]
```diff
@@ -127,11 +130,32 @@ const ModelServingPlatform: React.FC = () => {
...
       </DetailsSection>
       <ManageServingRuntimeModal
-        isOpen={isOpen}
+        isOpen={platformSelected === ServingRuntimePlatform.MULTI}
         currentProject={currentProject}
-        servingRuntimeTemplates={templatesEnabled}
+        servingRuntimeTemplates={templatesEnabled.filter((template) =>
+          getTemplateEnabledForPlatform(template, ServingRuntimePlatform.MULTI),
+        )}
+        onClose={(submit: boolean) => {
```
Comment (DaoDaoNoCode) [inline frontend/src/pages/modelServing/screens/projects/ModelServingPlatform.tsx:138]: This close function could be extracted above because it will be reused in both modals.

Comment (DaoDaoNoCode) [reply frontend/src/pages/modelServing/screens/projects/ModelServingPlatform.tsx:138]: Ah I mean just simply create a function
```
const onModalClose = (submit: boolean) => {
  setPlatformSelected(undefined);
  if (submit) {
    refreshAllProjects();
    refreshServingRuntime();
    refreshInferenceServices();
    setTimeout(refreshTokens, 500); // need a timeout to wait for tokens creation
  }
}
```
and apply it in both modal `onClose` prop `onClose={onModalClose}`

File: frontend/src/pages/modelServing/screens/projects/KServeSection/KServeInferenceServiceTableRow.tsx [hunk 3]
```diff
@@ -54,8 +64,28 @@ const KServeInferenceServiceTableRow: React.FC<KServeInferenceServiceTableRowPro
           servingRuntime={servingRuntimes.find(
             (sr) => sr.metadata.name === obj.spec.predictor.model.runtime,
           )}
-          onDeleteInferenceService={() => undefined}
-          onEditInferenceService={() => undefined}
+          onDeleteInferenceService={() => {
+            const servingRuntime = servingRuntimes.find(
+              (sr) => sr.metadata.name === obj.spec.predictor.model.runtime,
+            );
```
[the remark this answers is not in the document]
Comment (lucferbux) [reply frontend/src/pages/modelServing/screens/projects/KServeSection/KServeInferenceServiceTableRow.tsx:70]: yeah, that's what happens when you are refactoring in a rush, I'm dumb 🤦‍♂️

File: frontend/src/pages/modelServing/screens/projects/kServeModal/DeleteServingKServeModal.tsx [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+      onDelete={() => {
+        if (servingRuntime && inferenceService) {
+          setIsDeleting(true);
+
+          const { serviceAccountName, roleBindingName } = getTokenNames(
+            servingRuntime.metadata.name,
+            servingRuntime.metadata.namespace,
+          );
+
+          Promise.allSettled<ServingRuntimeKind | K8sStatus>([
```
[the remark this answers is not in the document]
Comment (lucferbux) [reply frontend/src/pages/modelServing/screens/projects/kServeModal/DeleteServingKServeModal.tsx:50]: yes, again, I think i refactored this and copied an  allsettled thinking about promise all, not sure why I want a granular control with allsettled if I'm doing a batch operation

File: frontend/src/pages/modelServing/screens/projects/kServeModal/DeleteServingKServeModal.tsx [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+            deleteServingRuntime(servingRuntime.metadata.name, servingRuntime.metadata.namespace),
+            deleteServiceAccount(serviceAccountName, servingRuntime.metadata.namespace),
+            deleteRoleBinding(roleBindingName, servingRuntime.metadata.namespace),
+            deleteInferenceService(
+              inferenceService.metadata.name,
+              inferenceService.metadata.namespace,
+            ),
+            ...tokens.map((token) => deleteSecret(token.metadata.namespace, token.metadata.name)),
```
[the remark this answers is not in the document]
Comment (lucferbux) [reply frontend/src/pages/modelServing/screens/projects/kServeModal/DeleteServingKServeModal.tsx:58]: ok, wasn't aware of that, let me try it

[the remark this answers is not in the document]
Comment (lucferbux) [reply frontend/src/pages/modelServing/screens/projects/kServeModal/DeleteServingKServeModal.tsx:58]: Yes, you are right!

File: frontend/src/pages/modelServing/screens/projects/KServeSection/KServeInferenceServiceTable.tsx [hunk 6]
```diff
@@ -1,28 +1,105 @@
...
+          },
+          inferenceServiceEditInfo: editKserveResources?.inferenceService,
+        }}
+        onClose={(submit: boolean) => {
+          setEditKServeResources(undefined);
+          if (submit) {
+            refreshServingRuntime();
+            refreshInferenceServices();
+            refreshDataConnections();
+            setTimeout(refreshTokens, 500); // need a timeout to wait for tokens creation
```
[the remark this answers is not in the document]
Comment (lucferbux) [reply frontend/src/pages/modelServing/screens/projects/KServeSection/KServeInferenceServiceTable.tsx:98]: Ok, I'm creating a follow up task to investigate if we can remove this.

File: frontend/src/pages/modelServing/screens/projects/kServeModal/ManageKServeModal.tsx [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+      ),
+      submitInferenceServiceResource(
+        createDataInferenceService,
+        setErrorModal,
+        onSuccess,
+        editInfo?.inferenceServiceEditInfo,
+        servingRuntimeName,
+        false,
+      ),
+    ]);
```
[the remark this answers is not in the document]
Comment (DaoDaoNoCode) [reply frontend/src/pages/modelServing/screens/projects/kServeModal/ManageKServeModal.tsx]: That's what I am experiencing. I guess I met some problems when deploy a KServe model. However, I saw an error message flickering very quickly and the modal closed. Then I cannot see anything in the table. I went to the cluster and found the model server (ServingRuntime) is already created. However, the model (InferenceService) is not created at all.

File: frontend/src/pages/modelServing/screens/global/InferenceServiceTable.tsx [hunk 8]
```diff
@@ -69,6 +70,13 @@ const InferenceServiceTable: React.FC<InferenceServiceTableProps> = ({
       <DeleteInferenceServiceModal
         isOpen={!!deleteInferenceService}
         inferenceService={deleteInferenceService}
+        servingRuntime={
+          deleteInferenceService && isModelMesh(deleteInferenceService)
+            ? undefined
+            : servingRuntimes.find(
+                (sr) => sr.metadata.name === deleteInferenceService?.spec.predictor.model.runtime,
+              )
```
Comment (christianvogt) [inline frontend/src/pages/modelServing/screens/global/InferenceServiceTable.tsx:78]:
```suggestion
          deleteInferenceService && !isModelMesh(deleteInferenceService)
            ? servingRuntimes.find(
                (sr) => sr.metadata.name === deleteInferenceService.spec.predictor.model.runtime,
              )
            : undefined
```

File: frontend/src/pages/modelServing/screens/global/useRouteForInferenceService.ts [hunk 9]
```diff
@@ -14,6 +16,13 @@ const useRouteForInferenceService = (
   const routeNamespace = inferenceService.metadata.namespace;
 
   React.useEffect(() => {
+    if (isKServe) {
+      const kserveRoute = getUrlFromKserveInferenceService(inferenceService);
+      setRoute(kserveRoute || null);
+      setLoaded(true);
+      setLoadError(kserveRoute ? null : new Error('Route not found'));
+      return;
+    }
```
Comment (christianvogt) [inline frontend/src/pages/modelServing/screens/global/useRouteForInferenceService.ts:25]: `inferenceService` is a bad object to use as a dependency on `useEffect` as unrelated changes can re-trigger the `useEffect`. Pre-compute the kserve route URL and use it as a dependency. For example:
```suggestion
  const kserveRoute = isKServe ? getUrlFromKserveInferenceService(inferenceService) : null;
  React.useEffect(() => {
    if (isKServe) {
      setRoute(kserveRoute);
      setLoaded(true);
      setLoadError(kserveRoute ? null : new Error('Route not found'));
      return;
    }
```

Comment (lucferbux) [reply frontend/src/pages/modelServing/screens/global/useRouteForInferenceService.ts:25]: oh right, ugh im sorry, that's a n00b error, thanks!
